web_src/js/features/comp/EditorMarkdown.js

```diff
@@ -0,0 +1,100 @@
+
+  for (const i of selectedLines) {
+    if (e.shiftKey) {
+      lines[i] = lines[i].replace(/^(\t| {1,2})/, '');
+    } else {
+      lines[i] = `  ${lines[i]}`;
+    }
+  }
+
+  // re-calculating the selection range
+  let newSelStart, newSelEnd;
+  pos = 0;
+  for (let i = 0; i < lines.length; i++) {
+    if (i === selectedLines[0]) {
+      newSelStart = pos;
+    }
+    if (i === selectedLines[selectedLines.length - 1]) {
+      newSelEnd = pos + lines[i].length;
+      break;
+    }
+    pos += lines[i].length + 1;
+  }
+  textarea.value = lines.join('\n');
+  textarea.setSelectionRange(newSelStart, newSelEnd);
+  triggerEditorContentChanged(textarea);
+}
+
+function handleNewline(textarea, e) {
+  const selStart = textarea.selectionStart;
+  const selEnd = textarea.selectionEnd;
+  if (selEnd !== selStart) return; // do not process when there is a selection
+
+  const value = textarea.value;
+
+  // find the current line
+  const lineStart = value.lastIndexOf('\n', selStart - 1) + 1;
+  let lineEnd = value.indexOf('\n', selStart);
+  lineEnd = lineEnd < 0 ? value.length : lineEnd;
+  let line = value.slice(lineStart, lineEnd);
+  if (!line) return; // if the line is empty, do nothing, let the browser handle it
+
+  // parse the indention
+  const indention = /^\s*/.exec(line)[0];
+  line = line.slice(indention.length);
+
+  // parse the prefixes: "1. ", "- ", "* ", "[ ] ", "[x] "
+  const prefixMatch = /^([0-9]+\.|[-*]|\[ \]|\[x\])\s/.exec(line);
+  let prefix = '';
+  if (prefixMatch) {
+    prefix = prefixMatch[0];
+    if (lineStart + prefix.length > selStart) prefix = ''; // do not add new line if cursor is at prefix
+  }
+
+  line = line.slice(prefix.length);
+  if (!indention && !prefix) return; // if no indention and no prefix, do nothing, let the browser handle it
+
+  e.preventDefault();
+  if (!line) {
+    // clear current line
+    textarea.value = value.slice(0, lineStart) + value.slice(lineEnd);
+  } else {
+    // start a new line with the same indention and prefix
+    let newPrefix = prefix;
+    if (newPrefix === '[x]') newPrefix = '[ ]';
+    if (/^\d+\./.test(newPrefix)) newPrefix = `1. `; // a simple approach, otherwise it needs to parse the lines after the current line
+    const newLine = `\n${indention}${newPrefix}`;
+    textarea.value = value.slice(0, selStart) + newLine + value.slice(selEnd);
+    textarea.setSelectionRange(selStart + newLine.length, selStart + newLine.length);
+  }
+  triggerEditorContentChanged(textarea);
+}
+
+export function initTextareaMarkdown(textarea) {
+  textarea.addEventListener('keydown', (e) => {
+    if (e.key === 'Tab' && !e.ctrlKey && !e.metaKey && !e.altKey) {
+      // use Tab/Shift-Tab to indent/unindent the selected lines
+      handleIndentSelection(textarea, e);
+    } else if (e.key === 'Enter' && !e.shiftKey && !e.ctrlKey && !e.metaKey && !e.altKey) {
+      // use Enter to insert a new line with the same indention and prefix
+      handleNewline(textarea, e);
+    }
```

Comment:
```suggestion
    if (e.ctrlKey || e.metaKey || e.altKey)
      return;
    if (e.key === 'Tab') {
      // use Tab/Shift-Tab to indent/unindent the selected lines
      handleIndentSelection(textarea, e);
    } else if (e.key === 'Enter' && !e.shiftKey) {
      // use Enter to insert a new line with the same indention and prefix
      handleNewline(textarea, e);
    }
```

Reply:
Style nit on that:

```js
    if (e.ctrlKey || e.metaKey || e.altKey) return;

Reply:
I do not think so. In the future we might want to handle more comb keys, `e.ctrlKey || e.metaKey || e.altKey` can't be a general condition.

Reply:
Yes, in the future.
But so far, that's not the case.
We can still change it once the need arises.

(@silverwind, I updated the suggestion)

Reply:
I do not this change is a must or blocker.

Reply:
> Yes, in the future.
> But so far, that's not the case.
> We can still change it once the need arises.

Then the problem is: in the future the code needs to change a lot and may introduce side effect.

My code structure is clear enough and never needs new code to touch old logic.